fortitude/src/check.rs

```diff
@@ -245,14 +268,85 @@ fn get_files<S: AsRef<str>>(files_in: &Vec<PathBuf>, extensions: &[S]) -> Vec<Pa
 }
 
 /// Parse a file, check it for issues, and return the report.
+#[allow(clippy::too_many_arguments)]
 pub(crate) fn check_file(
     path_rules: &Vec<PathRuleEnum>,
     text_rules: &Vec<TextRuleEnum>,
     ast_entrypoints: &BTreeMap<&str, Vec<AstRuleEnum>>,
     path: &Path,
     file: &SourceFile,
     settings: &Settings,
-) -> anyhow::Result<Vec<Diagnostic>> {
+    fix_mode: FixMode,
+    unsafe_fixes: UnsafeFixes,
+) -> anyhow::Result<Diagnostics> {
+    let (messages, fixed) = if matches!(fix_mode, FixMode::Apply | FixMode::Diff) {
+        if let Ok(FixerResult {
+            result,
+            transformed,
+            fixed,
+        }) = check_and_fix_file(
+            path_rules,
+            text_rules,
+            ast_entrypoints,
+            path,
+            file,
+            settings,
+            unsafe_fixes,
+        ) {
+            if !fixed.is_empty() {
+                match fix_mode {
+                    FixMode::Apply => {
+                        let mut out_file = File::create(path)?;
+                        out_file.write_all(transformed.source_text().as_bytes())?;
+                    }
+                    // TODO: diff
+                    FixMode::Diff => {}
+                    FixMode::Generate => {}
+                }
+            }
+
+            (result, fixed)
+        } else {
+            // Failed to fix, so just lint the original source
+            let result = check_only_file(
+                path_rules,
+                text_rules,
+                ast_entrypoints,
+                path,
+                file,
+                settings,
+            )?;
+            let fixed = FxHashMap::default();
+            (result, fixed)
+        }
+    } else {
+        let result = check_only_file(
+            path_rules,
+            text_rules,
+            ast_entrypoints,
+            path,
+            file,
+            settings,
+        )?;
+        let fixed = FxHashMap::default();
+        (result, fixed)
+    };
```

Comment:
We have this structure (in psuedo code):

```
if should_fix {
  try_fix
  if try_fix_worked {
     apply_fixes
  } else {
     just_lint
  }
} else {
  just_lint
}

return result
```

The first call to `check_only_file` here is a fall back if we tried to fix things but it didn't work for whatever reason.

The second call is if the user never requested fixes.

I think it could be rearranged, but we still end up with some repeated code:

```
if should_fix {
  try_fix
  if try_fix_worked {
     return result
  }
}

just_lint

return result
```

Ruff can't do this because it does some caching of the results before returning. If we intend to do that too, we'd have to switch back to the current structure

Reply:
Okay, that's fair. I think the problem looked bigger than it actually was because rustfmt had put everything on its own line.